### ISSUE TITLE
shorthands for var and const declarations, and infer types

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#1289]
This is for post-MVP.

We have a principle that expressions be (mostly) strongly typed.
We are also a bit hung up on possibly renaming 'const' and introducing 'constexpr' for closely related concepts. See #1272

I have been hoping to reduce wordiness along the following lines, for code inside a function body:

Before:
    
       var a : i32 = 12;
       var b : i32 = 13;
       const expr  : i32 = 1+2;
       const acopy : i32 = a;                //   Gets the value stored in a.
       const aptr : ptr<function,i32> =  a;  //   Gets the the pointer to a's storage.

After:

       var a: i32  = 12;   // same as before
       var b: = 13;        // type inference
       expr := 1+2;        // introduces new name 'expr' for the evaluation of 1+2.  Infer type.
       acopy := a;         // gets the value stored in a.
       aptr  ~= a;         // gets reference to the storage for a. Think of ~ as rope "tying" you to the storage.

Note that expr, acopy, aptr, are still 'const'.  They are not *re*-assignable.

Also, variables can't store pointers, so there is no desire/need/use-case for    `var a ~= ...`

**Comment by mehmetoguzderin** [comment]
> `       var b: = 13;        // type inference`

Is the `:` in `var b: = 13;` accidental? If not, I think it would be better to drop it.

**Comment by dneto0** [comment]
> Is the : in var b: = 13; accidental? If not, I think it would be better to drop it.

Good point.  I had not thought of that.  We already have 'var' to distinguish  the cases.
Either of these would be good:

          var b = 13;   
          var b := 13;   // This is more parallel to the type inference in const acopy := ....  But it's redundant like you say.

**Comment by Kangz** [comment]
`var b = 13;` would be nice. If we go the `:=` route to define consts, what do you think of allowing shadowing (only in the same scope, not child scopes) like so:

```
accumulator := initialValue;
accumulator := accumulator + computeSomeStuff();
accumulator := accumulator + computeMoreStuff();
```

Essentially each line would produce a new SSA value and tie it to the name `accumulator`.

**Comment by dj2** [comment]
That doesn't smell like a const if you're changing it ....

**Comment by Kangz** [comment]
It's a new SSA value every time, although I agree it would be more clear written as:

```
const accumulator := initialValue;
const accumulator := accumulator + computeSomeStuff();
const accumulator := accumulator + computeMoreStuff();
```

**Comment by dj2** [comment]
We don't expose the idea of SSA ID anywhere. It just feels weird that you're saying it's const and then changing the value.  Why not just do:

```
var a = initialValue;
a = a + computeSomeStuff();
const accumulator = a + computeMoreStuff();
```

**Comment by Kangz** [comment]
Because it would encourage people using `const` for everything except when they need explicit mutability. This is similar to the [`let` statement](https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html) (without `mut`) in Rust. Maybe @kvark has more experience of how useful it is in practice.

**Comment by dj2** [comment]
But ... it isn't const as you're changing it. That's my confusion here, const doesn't change, it just seems like a really weird thing to do to say it's const but sure you can change it go ahead, just put this magic word first.

The final `accumulator` mark that const as it doesn't change, but the part you want to change, that's not const.

**Comment by Kangz** [comment]
But it doesn't change, it's a new variable every time, it just shadows the previous one.

**Comment by dneto0** [comment]
The shadowing looks weird to me.

I'd just as soon write:
 
```
part1 := initialValue;
part2 :=  computeSomeStuff();
part3 := computeMoreStuff();
sum := part1 + part2 + part3;
```

**Comment by jdashg** [comment]
I love the idea of allowing en-consting shadowing of variables, but it's (mostly?) orthogonal to this shorthand.

**Comment by jdashg** [comment]
For shorthands, I think we get the most bang-for-the-buck from type inference, which we should just do. (including inference of the quite-verbose pointer types, maybe via `const aptr = &a;`) Other shorthands have more diminishing returns, and I think might not be worth it vs the extra syntax/glyphs we impose on code-readers.

**Comment by kainino0x** [comment]
I personally don't want to have a new syntax (`:=`) that creates an identifier without a keyword in front. Better to be consistent with what we have.

Also agree I think we'd get all of the convenience by having type inference on var and const.

For pointers, I prefer the pointer operator `&x` over the pointer assignment `~= x`. As an aside, do we really allow `const aptr : ptr<function,i32> =  a;` right now? That goes against the "one expression one type" rule we agreed upon (`a` can be treated as either `i32` OR `ptr<function, i32>`). Sorry, I'm not very clear on what our pointer story is right now.

**Comment by Kangz** [comment]
Closing. Types can be omitted in `const` and `let` declarations (at least in function scope).

**Comment by jdashg** [comment]
This is specifically for proposing an even-shorter shorthand, not just inference.
